### PR TITLE
Add BlackForest Flux Support 

### LINKED ISSUE #763: Support for Flux model in diffusers
### Feature request

Hey @yahavb, to continue our discussion on supporting the Flux model here. 

### Add export support for each component

* Text encoder 1

No need to add the wrapper, it should be already handled by `CLIPTextNeuronConfig`.

* Text encoder 2

Can reuse the `T5EncoderForDiffusersNeuronConfig` class, in which the `T5EncoderWrapper` is equivalent to `TracingT5TextEncoderWrapper`. 
For tp support, will need to enable `tensor_parallel_size` arg for many functions in `optimum/exporters/neuron/__main__.py` (eg. `_get_submodels_and_neuron_configs_for_stable_diffusion`). And we need to write a `get_parallel_callable` for the neuron config class to shard the model, we can take the one in `T5EncoderForTransformersNeuronConfig` as a reference.

* Transformer

In the [script](https://github.com/yahavb/aws-neuron-eks-samples/blob/master/flux_serve/app/src/transformer/compile.py), transformer_embedders / transformer_blocks / ingle_transformer_blocks / transformer_out_layers are traced separately, would it be possible to fit into neuron devices while tracing the whole transformer together? If it's not the case, in Optimum Neuron logic, each traced component means a NeuronConfig class, then you will need to create a class for each component and build a function to fetch each module (similar to `_get_submodels_and_neuron_configs_for_stable_diffusion`), otherwise, you would be able to create a single `FluxTransformerNeuronConfig`, and add the tp config there.

* VAE decoder

Nothing to do, already supported

### Test

Here is an example of testing the compilation:

```bash
optimum-cli export neuron --model black-forest-labs/FLUX.1-dev --tensor_parallel_size 8 --batch_size 1 --height 1024 --width 1024 --num_images_per_prompt 1 --sequence_length 512 --torch_dtype bfloat16 flux_neuron/
```

## 2nd step: Inference support

Depends on the 1st step to decide if we need to override some functions in `FluxPipeline` (if we traced transformer in multiple module), but whatever we need to put under `optimum/neuron/pipelines/diffusers`. We will need to add the following class under `optimum/neuron/modeling_diffusion.py`:

```python
class NeuronFluxPipeline(NeuronDiffusionPipelineBase, FluxPipeline):
    main_input_name = "prompt"
    auto_model_class = FluxPipeline
```

Reference PRs:

* Add TP support for T5 model: #697
* Add Pixart support: #762 

### Motivation

m

### Your contribution

,

### PR DESCRIPTION
**Pull request by JingyaHuang**
# What does this PR do?

Fixes #763,  #676 

## Compilation

- [x] Export of Flux pipeline (`TP=8` for Flux transformer 2D) with `neuronx_distributed.trace.model_builder.ModelBuilder`

### Export via CLI

* Regular 

```bash
optimum-cli export neuron --model black-forest-labs/FLUX.1-dev --tensor_parallel_size 8 --batch_size 1 --height 768 --width 1360 --num_images_per_prompt 1 --torch_dtype bfloat16 flux_neuron/
```

* Tiny test

```bash
optimum-cli export neuron --model hf-internal-testing/tiny-flux-pipe-gated-silu --tensor_parallel_size 2 --batch_size 1 --height 8 --width 8 --num_images_per_prompt 1 --sequence_length 256 --torch_dtype bfloat16 tiny_flux_neuron/
```

### Export with `NeuronFluxPipeline` API

```python
from optimum.neuron import NeuronFluxPipeline

if __name__ == "__main__":
    compiler_args = {"auto_cast": "none"}
    input_shapes = {"batch_size": 1, "height": 1024, "width": 1024}

    pipe = NeuronFluxPipeline.from_pretrained(
        "black-forest-labs/FLUX.1-dev",
        torch_dtype=torch.bfloat16,
        export=True,
        tensor_parallel_size=8,
        # disable_neuron_cache=True,
        **compiler_args,
        **input_shapes
    )

    # Save locally
    pipe.save_pretrained("flux_dev_neuron_1024_tp8/")

    # Upload to the HuggingFace Hub
    pipe.push_to_hub(
        "flux_dev_neuron_1024_tp8/", repository_id="Jingya/FLUX.1-dev-neuronx-1024x1024-tp8"  # Replace with your HF Hub repo id
    )

```

You can find an example of compiled artifacts [here (`Jingya/flux.1-dev_neuronx_tp8` )](https://huggingface.co/Jingya/flux.1-dev_neuronx_tp8/tree/main)

## Inference

- [x] Flux Inference

For generating an image with `NeuronFluxPipeline`:

```python
from optimum.neuron import NeuronFluxPipeline

pipe = NeuronFluxPipeline.from_pretrained("flux_neuron")
prompt = "A cat holding a sign that says hello world"
out = pipe(
    prompt,
    guidance_scale=3.5,
    num_inference_steps=50,
    generator=torch.Generator("cpu").manual_seed(0)
).images[0]
out.save("flux_optimum.png")
```

## Other
- [x] Tests
- [x] Doc

cc. @yahavb

### CODE REVIEW

File: optimum/exporters/neuron/convert.py [hunk 1]
```diff
@@ -570,34 +626,72 @@ def export_neuronx(
...
+            model_builder.add(
+                key=subfolder,
+                model_instance=model,
+                example_inputs=[dummy_inputs],
+                priority_model_idx=0,
+                compiler_args=compiler_args,
+            )
+            neuron_model = model_builder.trace(initialize_model_weights=False)
+
+            model_builder.shard_checkpoint(serialize_path=output.parent / "weights/")
```
Comment (dacorvo) [inline optimum/exporters/neuron/convert.py:675]: Eventually this could be omitted: the weights can be sharded at loading time only, which makes export a lot faster. You only need to remember where the weights are (local dir or hub).

File: optimum/neuron/models/inference/flux/modeling_flux.py [hunk 2]
```diff
@@ -0,0 +1,904 @@
...
+        )
+        self.proj_out = ColumnParallelLinear(
+            self.inner_dim,
+            patch_size * patch_size * self.out_channels,
+            bias=True,
+            gather_output=True,
+            reduce_dtype=reduce_dtype,
+        )
+
+    def forward(
```
Comment (dacorvo) [inline optimum/neuron/models/inference/flux/modeling_flux.py:206]: It does not look like this method has anything neuron specific: wouldn't have been possible to inherit from the `diffusers` class and just override `__init__` ?

File: tests/inference/test_nxd.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
+# coding=utf-8
```
Comment (dacorvo) [inline tests/inference/test_nxd.py]: You should rename the test file to make it flux specific, like `test_flux_pipeline.py`.

File: tests/inference/test_nxd.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    NeuronModelVaeEncoder,
+)
+from optimum.neuron.utils.testing_utils import is_inferentia_test, requires_neuronx
+from optimum.utils.testing_utils import require_diffusers
+
+
+@is_inferentia_test
+@requires_neuronx
+@require_diffusers
+class NeuronFluxPipelineIntegrationTest(unittest.TestCase):
```
Comment (dacorvo) [inline tests/inference/test_nxd.py]: Is it required to wrap this in a `unittest.TestCase` ? It looks like it could be a simple test function, which is easier to test individually and parametrize with `pytest`.

File: docs/source/model_doc/diffusers/flux.mdx [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+# Flux
+
+Flux is a series of text-to-image generation models based on diffusion transformers.
+
+> [!TIP]
+> We recommend using a `inf2.24xlarge` instance with tensor parallel size 8 for the model compilation and inference.
```
Comment (tengomucho) [inline docs/source/model_doc/diffusers/flux.mdx:22]: why 24x if we only do TP8?

File: tests/inference/test_nxd.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+            **input_shapes,
+            **compiler_args,
+        )
+        self.assertIsInstance(neuron_pipeline.text_encoder, NeuronModelTextEncoder)
+        self.assertIsInstance(neuron_pipeline.text_encoder_2, NeuronModelTextEncoder)
+        self.assertIsInstance(neuron_pipeline.transformer, NeuronModelTransformer)
+        self.assertIsInstance(neuron_pipeline.vae_encoder, NeuronModelVaeEncoder)
+        self.assertIsInstance(neuron_pipeline.vae_decoder, NeuronModelVaeDecoder)
+
+        prompt = "Mario eating hamburgers."
```
Comment (dacorvo) [inline tests/inference/test_nxd.py]: Another food-based prompt 🤔 ... is it your signature 😉 ?

**Review by dacorvo — CHANGES_REQUESTED** [review]
Massive pull-request: impressive !!!! I only have a few minor comments.
Eventually, we should think about how some of the `ModelBuilder` related code can be reused for non-decoder models: it is unclear to me yet if there are some specifics that would prevent that though.

File: optimum/exporters/neuron/base.py [hunk 7]
```diff
@@ -380,17 +382,24 @@ def unflatten_inputs(cls, inputs: dict[str, Any]) -> dict[str, Any]:
...
-        dummy_inputs: dict[str, torch.Tensor] | None = None,
+        input_names: list[str] = None,
         forward_with_tuple: bool = False,
         eligible_outputs: list[str | int] | None = None,
         device: str | None = None,
     ):
         """
-        Checks if inputs order of the model's forward pass correspond to the generated dummy inputs to ensure the dummy inputs tuple used for
-        tracing are under the correct order.
+        Patch the model and generate aliased for tracing.
```
Comment (tengomucho) [inline optimum/exporters/neuron/base.py]:
```suggestion
        Patch the model and generate aliases for tracing.
```

File: optimum/exporters/neuron/convert.py [hunk 8]
```diff
@@ -529,20 +532,68 @@ def export_neuronx(
...
+        dummy_inputs_tuple = tuple(dummy_inputs.values())
+        return dummy_inputs_tuple
+
+
+def prepare_compiler_flags(
+    config: "NeuronDefaultConfig",
+    auto_cast: str | None = None,
+    auto_cast_type: str = "bf16",
+    optlevel: str = "2",
+    instance_type: str = "trn1",
```
Comment (tengomucho) [inline optimum/exporters/neuron/convert.py:595]: I thought you suggested to use inf2

File: optimum/exporters/neuron/utils.py [hunk 9]
```diff
@@ -436,42 +471,47 @@ def get_submodels_for_export_diffusion(
...
-            if attn_mask is not None:
-                return sdpa_original(query, key, value, attn_mask=attn_mask, dropout_p=dropout_p, is_causal=is_causal)
-            else:
-                return neuron_scaled_dot_product_attention(
-                    query, key, value, attn_mask=attn_mask, dropout_p=dropout_p, is_causal=is_causal
-                )
-
-        torch.nn.functional.scaled_dot_product_attention = attention_wrapper
-        models_for_export.append((DIFFUSION_MODEL_TRANSFORMER_NAME, copy.deepcopy(transformer)))
+        if not is_flux:  # The following will be handled by `ModelBuilder` if `is_flux`.
```
Comment (tengomucho) [inline optimum/exporters/neuron/utils.py:480]: since you do not have an "else" here, maybe you can combine the two if and reduce indentation.

File: optimum/exporters/neuron/convert.py (hunk 1 above)
Comment (JingyaHuang) [reply optimum/exporters/neuron/convert.py:675]: Do you mean we shard the weights during the loading time instead of after the tracing? Indeed, it increases the whole export time, but somehow, I would rather spend more time exporting one shot and have faster loading + warmup during the deployment.

**Review by tengomucho — APPROVED** [review]
Few nits, otherwise LGTM!

File: optimum/neuron/models/inference/flux/modeling_flux.py (hunk 2 above)
Comment (JingyaHuang) [reply optimum/neuron/models/inference/flux/modeling_flux.py:206]: here I change the inputs from `txt_ids` and `img_ids` in diffusers to `image_rotary_emb` that we will compute during the inference runtime on CPU.

File: optimum/exporters/neuron/convert.py (hunk 1 above)
Comment (dacorvo) [reply optimum/exporters/neuron/convert.py:675]: Maybe in your case it is not that bad, because sharding does not mean loading the weights on the device, does it ?
It is just that you don't cache the sharded weights, so it is only useful when you want to push the exported model to the hub.
This is what I do for decoders:
- when using the optimum-cli: export (cache or fetch NEFFs)
- when using from_pretrained: export (cache or fetch NEFFs)  + load_weights
For large models like llama 70B this makes a huge difference as loading weights takes several minutes.

File: tests/inference/test_nxd.py (hunk 4 above)
Comment (JingyaHuang) [reply tests/inference/test_nxd.py]: no, pytest is much better. Let me change it.

File: tests/inference/test_nxd.py (hunk 6 above)
Comment (JingyaHuang) [reply tests/inference/test_nxd.py]: doing so much physio these days, need energy from these food prompts...

File: docs/source/model_doc/diffusers/flux.mdx (hunk 5 above)
Comment (JingyaHuang) [reply docs/source/model_doc/diffusers/flux.mdx:22]: Because inf2 jump from either you want to have just 1 neuron device, or you want to have 6
image

File: optimum/exporters/neuron/convert.py (hunk 8 above)
Comment (JingyaHuang) [reply optimum/exporters/neuron/convert.py:595]: interestingly the compiler flag we put for inf2 export is "trn1"

Comment (JingyaHuang) [reply optimum/exporters/neuron/convert.py:595]: don't ask why

File: optimum/exporters/neuron/utils.py (hunk 9 above)
Comment (JingyaHuang) [reply optimum/exporters/neuron/utils.py:480]: yeah but for pixart (case diffusion transformer but not flux) I would still need to apply the following:

https://github.com/huggingface/optimum-neuron/blob/7746b1c27fefe87b95e667705a3d1c16cf3e1670/optimum/exporters/neuron/utils.py#L497-498

File: docs/source/model_doc/diffusers/flux.mdx (hunk 5 above)
Comment (tengomucho) [reply docs/source/model_doc/diffusers/flux.mdx:22]: oh, right

File: optimum/exporters/neuron/convert.py (hunk 1 above)
Comment (JingyaHuang) [reply optimum/exporters/neuron/convert.py:675]: na na, `shard_checkpoint` just shard the weights, then we either serialize things to disk like I do here or load to Neuron.

Comment (JingyaHuang) [reply optimum/exporters/neuron/convert.py:675]: Yeah, I should definitly look into the neff cache part of ModelBuilder (not yet the case) and the work you have done that I could reuse. will do it next!

**Review by dacorvo — APPROVED** [review]
LGTM, thanks !

**Review by tengomucho — APPROVED** [review]
